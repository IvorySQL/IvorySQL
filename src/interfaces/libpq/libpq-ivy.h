/*-------------------------------------------------------------------------
 *
 * libpq-ivy.h like libpq-fe.h but contains some others ivy functions
 * This file contains definitions for structures and
 * externs for functions used by frontend applications.
 *
 * Portions Copyright (c) 2025, IvorySQL Global Development Team
 *
 * src/interfaces/libpq/libpq-ivy.h
 *
 *-------------------------------------------------------------------------
 */

#ifndef LIBPQ_IVY_H
#define LIBPQ_IVY_H

#ifdef __cplusplus
extern		"C"
{
#endif

#include <stdio.h>
#include "pg_config.h"
#include "ivy_sema.h"
#include "postgres_ext.h"

/* argument modes */
typedef enum
{
	argmode_in,
	argmode_out,
	argmode_inout
} Ivyargmode;

typedef struct Ivylist
{
	void *value;
	struct Ivylist *next;
} Ivylist;


/* bind out parameter structure */
typedef struct IvyBindOutInfo
{
	struct IvyBindOutInfo *next;
	int 	position;		/* variable position */
	int	resultcolumnno;		/* the column number in the PGresult */
	void 	*var;			/* the address to save out parameters return value */
	int 	*indp;			/* index the value copy status */
	size_t 	val_size;	/* the size of var */
	int	dtype;		/* the type of parameter */

	/*
	 * these parameter come from OCI,
	 * and may be used in further.
	 */
	char *alenp;
	char *recodep;
	int  maxarr_len;
	int  *curelep;
	int  mode;
} IvyBindOutInfo;

typedef struct IvyBindOutInfo IvyBindInfo;

typedef enum IvyStmtType
{
	IVY_STMT_UNKNOW,
	IVY_STMT_DO,
	IVY_STMT_DOHANDLED,
	IVY_STMT_OTHERS
} IvyStmtType;

typedef struct IvyBindOutNameInfo
{
	IvyBindOutInfo 		**bindinfo;
	void			*var;			/* bindvar address */
	int			*indp;
	size_t			val_size;
	char			*name;			/* placeholdvar name */
	size_t			name_len;		/* placeholdvar name len */
	int			replace;		/* has replace to position */
	int			dtype;			/* the type of parameter */
	/*
	 * these parameter come from OCI, 
	 * and maybe used in further.
	 */
	char *alenp;
	char *recodep;
	int  maxarr_len;
	int  *curelep;
	int  mode;
	struct IvyBindOutNameInfo *next;
} IvyBindOutNameInfo;

typedef struct IvyBindOutNameInfo IvyBindNameInfo;

typedef struct
{
	PGconn *conn;
	PGSemaphoreData self_lock;	/* conn modification need this lock */
	PGSemaphoreData lock;		/* IvystmtHandleList modification need this lock */
	Ivylist *IvystmtHandleList;
} Ivyconn;

/* like PGresult */
typedef struct
{
	PGresult *result;
	int	off;		/* out parameter offset */
} Ivyresult;

/* similar to jdbc statement handle */
typedef struct
{
	char *stmtName;
	char *query;
	int query_len;
	int nParams;
	Oid *paramTypes;
	int	mode;
	char	**paramNames;	/* save param names */
	int	language;	

	PGSemaphoreData lock;	/* datamodification need this lock */
	Ivylist *IvyconnList;
	int	noutvars;
	IvyBindOutInfo *outbind;
	IvyBindOutNameInfo *namebind;	/* binded by name */
	int		name_replace;		/* by name changed to by position */
	IvyStmtType	stmttype;		/* stmt type */
	char		*do_using_query;	/* Dostmt should rewrite query */
} IvyPreparedStatement;

typedef struct IvyError
{
	size_t err_buf_size;
	char	*error_msg;
} IvyError;

typedef enum IVY_HANDLE_TYPE
{
	IVY_HANDLE_STMT,
	IVY_HANDLE_ERROR,
	IVY_HANDLE_NO
} ivy_handle_type;

typedef struct HostVariableEntry
{
	char	*name;
	int		position;
} HostVariableEntry;

typedef struct HostVariable
{
	HostVariableEntry *hostvars;
	int		length;
	bool	isdostmt;
} HostVariable;

extern Ivyconn *Ivyconnectdb(const char *conninfo);
extern Ivyconn *Ivyconnectdb2(PGconn *conn);
extern void Ivyfinish2(Ivyconn *conn);
extern IvyPreparedStatement *IvyCreatePreparedStatement(const char *stmtName,
								const char *query,
								int nParams,
								const Oid *paramTypes);
extern int IvybindOutParameterByPos(IvyPreparedStatement *stmtHandle,
					IvyBindOutInfo **bindinfo,
					int position,
					void *var,
					size_t	val_size,
					int *indp,
					int replace_ok,
					char *errmsg,
					size_t size_errmsg);
extern int IvyHandleAlloc(const void *parent,
				void **hndlpp,
				ivy_handle_type	type,
				size_t	size,
				void	**usermapp);
extern int IvyStmtPrepare(IvyPreparedStatement * stmt,
				IvyError *errhp,
				const char *query,
				size_t	query_len,
				int language,
				int mode);
extern void IvyFreeHandle(void *handle, ivy_handle_type type);
extern int IvyBindByPos(IvyPreparedStatement *stmtHandle,
				IvyBindInfo **bindinfo,
				IvyError *errhp,
				int position,
				void *valuep,
				size_t	value_sz,
				int dty,
				int *indp,
				char *alenp,
				char *recodep,
				int  maxarr_len,
				int  *curelep,
				int  mode);
extern int IvyBindByName(IvyPreparedStatement *stmtHandle,
				IvyBindInfo **bindinfo,
				IvyError *errhp,
				const char *name,
				size_t name_len,
				void *var,
				size_t	val_size,
				int dty,
				int *indp,
				char *alenp,
				char *recodep,
				int  maxarr_len,
				int  *curelep,
				int  mode);
extern Ivyresult *IvyStmtExecute(Ivyconn *tconn,
					IvyPreparedStatement *stmtHandle,
					IvyError *errhp);

extern Ivyresult *IvyStmtExecute2(Ivyconn *tconn,
					  				IvyPreparedStatement *stmtHandle,
					  				IvyError *errhp,
					  				HostVariable *host);

extern int IvybindOutParameterByName(IvyPreparedStatement *stmtHandle,
					IvyBindOutInfo **bindinfo,
					const char *name,
					size_t name_len,
					void *var,
					size_t	val_size,
					int *indp,
					int replace_ok,
					char *errmsg,
					size_t size_errmsg);

extern Ivyresult *IvyexecPreparedStatement(Ivyconn *conn,
						IvyPreparedStatement *stmtHandle,
						int nParams,
						const char *const *paramValues,
						const int *paramLengths,
						const int *paramFormats,
						const Ivyargmode *argmodes,
						int resultFormat,
						char *errmsg,
						size_t size_errmsg);

extern Ivyresult *IvyexecPreparedStatement2(Ivyconn *conn,
						IvyPreparedStatement *stmtHandle,
						int nParams,
						const char *const *paramValues,
						const int *paramLengths,
						const int *paramFormats,
						const Ivyargmode *argmodes,
						IvyBindOutInfo *bindinfos,
						int resultFormat,
						char *errmsg,
						size_t size_errmsg);

extern void IvyFreePreparedStatement(IvyPreparedStatement *stmtHandle);
extern void Ivyclear(Ivyresult *tresult);
extern void Ivyfinish(Ivyconn *conn);

extern ConnStatusType Ivystatus(const Ivyconn *conn);
extern Ivyresult *Ivyexec(Ivyconn *conn, const char *query);
extern ExecStatusType IvyresultStatus(const Ivyresult *res);
extern int Ivynfields(const Ivyresult *tresult);
extern char *Ivyfname(const Ivyresult *res, int field_num);
extern int Ivyntuples(const Ivyresult *res);
extern char *Ivygetvalue(const Ivyresult *res, int tup_num, int field_num);
extern char *IvyerrorMessage(const Ivyconn *conn);

#ifdef __cplusplus
}
#endif

#endif   /* LIBPQ_IVY_H */

