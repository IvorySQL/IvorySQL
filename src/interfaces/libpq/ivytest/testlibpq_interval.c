/*-------------------------------------------------------------------------
 *
 * testlibpq_interval.c
 *
 * Test the C version of LIBPQ, the POSTGRES frontend library.  
 *
 * Portions Copyright (c) 2025, IvorySQL Global Development Team
 *
 * src/interfaces/libpq/ivytest/testlibpq_interval.c
 *
 *-------------------------------------------------------------------------
 */

#include <stdio.h>
#include <stdlib.h>
#include "libpq-fe.h"
#include "libpq-ivy.h"
#include <string.h>
#include <stdint.h>
#include <string.h>
#include <netinet/in.h>
#include <sys/types.h>

char *function_str1 = "create or replace function test_interval(id integer, time1 out interval year to month, time2 out interval year to month[], time3 interval year to month) return interval year to month as "
"  mds interval year to month[3];"
" begin"
"    mds[0] := interval '1-1' year to month;"
"    mds[1] := interval '1-2' year to month;"
"    mds[2] := interval '1-3' year to month;"
"    time1 := interval '1-2' year to month;"
"    time2 := mds;"
"    return time3;"
"end;";

char *function_str2 = "create or replace function test_interval(id integer, time1 out interval day to second, time2 out interval day to second[], time3 interval day to second) return interval day to second as "
"  mds interval day to second[3];"
" begin"
"    mds[0] := interval '4 13:12:12' day to second;"
"    mds[1] := interval '5 13:12:12' day to second;"
"    mds[2] := interval '6 13:12:12' day to second;"
"    time1 := interval '4 12:12:12' day to second;"
"    time2 := mds;"
"    return time3;"
"end;";


static void
exit_nicely(Ivyconn *conn)
{
	Ivyfinish(conn);
	exit(EXIT_SUCCESS);
}

static void
exec_prepare(Ivyconn *conn, int flag)
{
	int   nFields;
	int   i,j;
	int   result_format = 0;
	Ivyresult   *res;
	const char *stmt_name = "test_stmt";
	const char *stmt = "select * from test_interval($1,$2,$3,$4);";
	Oid param_types[4];
	IvyPreparedStatement *stmthandle;
	const char* custid = "200";
	const char* time1 = "'1-1'";
	const char* time2 = "'1-2'";
	const char* time3 = "'1-3'";
	const char* param_values[4];
	int param_lengths[4];
	int param_formats[4];
	char errmsg[256];
	IvyBindOutInfo *bindinfo[2] = {NULL, NULL};
	char bindvar1[256];
	char bindvar2[256];
	int index[2];
	Ivyargmode modes[4];

	memset(bindvar1, 0x00, 256);
	memset(bindvar2, 0x00, 256);

	if (flag == 1)
	{
		param_types[0] = 23 | 0x20000000  /* in */;
		param_types[1] = 9516 | 0x40000000  /* out */;
		param_types[2] = 9517 | 0x40000000  /* out */;
		param_types[3] = 9516 | 0x20000000  /* in */;
	}
	else
	{
		param_types[0] = 23 | 0x20000000  /* in */;
		param_types[1] = 9518 | 0x40000000  /* out */;
		param_types[2] = 9519 | 0x40000000  /* out */;
		param_types[3] = 9518 | 0x20000000  /* in */;
		time1 = "'4 12:12:12'";
		time2 = "'4 13:12:12'";
		time3 = "'4 14:12:12'";
	}
	stmthandle = IvyCreatePreparedStatement(stmt_name, stmt, 4, param_types);
	IvybindOutParameterByPos(stmthandle,
							&bindinfo[0],
							2,
							bindvar1,
							256,
							&index[0],
							1,
							errmsg,
							256);
	IvybindOutParameterByPos(stmthandle,
							&bindinfo[1],
							3,
							bindvar2,
							256,
							&index[1],
							1,
							errmsg,
							256);
	param_values[0] = custid;
	param_values[1] = time1;
	param_values[2] = time2;
	param_values[3] = time3;
	param_lengths[0] = strlen(custid);
	param_lengths[1] = strlen(time1);
	param_lengths[2] = strlen(time2);
	param_lengths[3] = strlen(time3);

	param_formats[0] = 0;
	param_formats[1] = 0;
	param_formats[2] = 0;
	param_formats[3] = 0;
	modes[0] = argmode_in;
	modes[1] = modes[2] = argmode_out;
	modes[3] = argmode_in;
	res = IvyexecPreparedStatement(conn, stmthandle, 4, param_values, param_lengths,
								param_formats, modes, result_format, errmsg, 256);
	if (IvyresultStatus(res) != PGRES_TUPLES_OK && IvyresultStatus(res) != PGRES_COMMAND_OK &&
		IvyresultStatus(res) != PGRES_EMPTY_QUERY)
	{
		fprintf(stderr, "PQexecPrepared statement didn't return tuples properly\n");
		Ivyclear(res);
		exit_nicely(conn);
	}
	nFields = Ivynfields(res);
	for (i = 0; i < nFields; i++)
		printf("%-15s", Ivyfname(res, i));

	printf("\n");
	/* print out the instances */
	for (i = 0; i < Ivyntuples(res); i++)
	{
		for (j = 0; j < nFields; j++)
		{
			printf("%-15s", Ivygetvalue(res, i, j));
			printf("\t");
		}
		printf("\n");
	}

	printf("%s,%s\n",bindvar1,bindvar2);
	printf("%d,%d\n",index[0],index[1]);
	Ivyclear(res);
	IvyFreePreparedStatement(stmthandle);
	return;
}

int
main()
{
	Ivyconn    *conn;
	Ivyresult   *res;
	const char *conninfo="user=system dbname=postgres port=1521";

	/* make a connection to the database */
	conn = Ivyconnectdb(conninfo);

	/* check to see that the backend connection was successfully made */
	if (Ivystatus(conn) == CONNECTION_BAD)
	{
		fprintf(stderr, "Connection to database failed.\n");
		fprintf(stderr, "%s", IvyerrorMessage(conn));
		exit_nicely(conn);
	}
	res = Ivyexec(conn, function_str1);
	Ivyclear(res);
	res = Ivyexec(conn,function_str2);
	Ivyclear(res);

	/* start a transaction block */
	res = Ivyexec(conn, "BEGIN");

	if (IvyresultStatus(res) != PGRES_COMMAND_OK)
	{
		fprintf(stderr, "BEGIN command failed\n");
		Ivyclear(res);
		exit_nicely(conn);
	}
	Ivyclear(res);

	exec_prepare(conn, 1);
	exec_prepare(conn, 2);
	/* end the transaction */
	res = Ivyexec(conn, "END");
	Ivyclear(res);

	res = Ivyexec(conn, "drop function test_interval(integer, interval year to month, interval year to month[], interval year to month);");
	Ivyclear(res);
	res = Ivyexec(conn, "drop function test_interval(integer, interval day to second, interval day to second[], interval day to second);");
	Ivyclear(res);

	Ivyfinish(conn);
	return 0;
}

