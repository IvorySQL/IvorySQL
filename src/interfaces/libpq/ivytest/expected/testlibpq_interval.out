_RETVAL_       
+000000001-03  	
+000000001-02,[0:2]={+000000001-01,+000000001-02,+000000001-03}
0,0
_RETVAL_       
+000000004 14:12:12.000000000	
+000000004 12:12:12.000000000,[0:2]={"+000000004 13:12:12.000000000","+000000005 13:12:12.000000000","+000000006 13:12:12.000000000"}
0,0
